pkg/ebpf/verifier/elf.go

```diff
@@ -0,0 +1,292 @@
+		return nil, nil, fmt.Errorf("cannot get line reader for %s: %w", file, err)
+	}
+
+	// Build the map that links each symbol to the sequence index it belongs to
+	symToSeq, err := buildSymbolToSequenceMap(elfFile)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot build symbol to section index map for %s: %w", file, err)
+	}
+
+	// Build the map that links the sequence index + address to the start of each program, as the DWARF
+	// line info data doesn't tell you which program a line belongs to.
+	progStartMap, err := buildProgStartMap(dwarfData, symToSeq)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot build program start lines for %s: %w", file, err)
+	}
+
+	// Now build the map that, for each program, links the instruction offset to line information
+	// Read all the lines in the .debug_lines section
+	offsets := make(map[string]map[uint64]string)
+	currProgram := ""
+	startingOffset := uint64(0)
+	sequenceIndex := 0
+	prevAddress := 0
+	for {
+		var line dwarf.LineEntry
+		err := lineReader.Next(&line)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, nil, fmt.Errorf("DWARF lineReader file %s: %w", file, err)
+		}
+
+		// Ignore lines with no data
+		if line.File == nil || line.Line == 0 {
+			continue
+		}
+
+		// Increase section indexes whenever we reset the program. We have to look at the value of the previous
+		// address, because we might have multiple lines that have the same zero address.
+		if line.Address == 0 && prevAddress != 0 {
+			sequenceIndex++
+		}
+		prevAddress = int(line.Address)
+
+		startPoint := progStartPoint{sequenceIndex, int64(line.Address)}
+		lineinfo := fmt.Sprintf("%s:%d", line.File.Name, line.Line)
+
+		// Reset the current program only if it's the first time we see it. Multiple
+		// assembly instructions might point to the first source line of a program
+		if newProg, ok := progStartMap[startPoint]; ok && newProg != currProgram {
+			// We need to keep track of the starting offset for each program to calculate
+			// the offset relative to the start. The eBPF loaders count program instructions
+			// from the start of the program, while in the ELF binary they're relative to the
+			// section start, and we might have multiple functions per section.
+			startingOffset = line.Address
+			currProgram = progStartMap[startPoint]
+		}
+
+		if currProgram == "" {
+			// We might have information of programs that are not in the spec, ignore those
+			continue
+		}
+		if _, ok := offsets[currProgram]; !ok {
+			offsets[currProgram] = make(map[uint64]string)
+		}
+		offset := line.Address - startingOffset
+		offsets[currProgram][offset] = lineinfo
+	}
+
+	// Now that we have line information for each instruction, we can build the source map
+	sourceMap := make(map[string]map[int]*SourceLine)
+	funcsPerSection := make(map[string][]string)
+	currLineInfo := ""
+	currLine := ""
+	for _, progSpec := range spec.Programs {
+		sourceMap[progSpec.Name] = make(map[int]*SourceLine)
```

Comment:
Is the range of keys continuous (`0-x`)? If so, would a `[]*SourceLine` slice make more sense?

Reply:
Not necessarily. Instructions can start at an arbitrary offset, and it's not continuous. The instruction index as used by the verifier and the DWARF data is the offset divided by 8, but some instructions (LdImmDW) has 16-bytes instead of 8, so whenever it's there the instruction index jumps from x to x + 2.